File: callouts/python/service_callout.py
# Copyright 2023 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""SDK for service callout servers.

Provides a customizeable, out of the box, service callout server.
Takes in service callout requests and performs header and body transformations.
Bundled with an optional health check server.
Can be set up to use ssl certificates.
"""
from concurrent import futures
from http.server import BaseHTTPRequestHandler, HTTPServer
from typing import AnyStr, Iterable, Iterator

import grpc
from grpc import ServicerContext
import service_pb2
import service_pb2_grpc


def add_header_mutation(
    add: list[tuple[str, str]] | None = None,
    remove: list[str] | None = None,
    clear_route_cache: bool = False,
) -> service_pb2.HeadersResponse:
  """Generate a header response for incoming requests.

  Args:
    add: A list of tuples representing headers to add.
    remove: List of header strings to remove from the request.
    clear_route_cache: If true, will enable clear_route_cache on the response.

  Returns:
    The constructed header response object.
  """
  header_mutation = service_pb2.HeadersResponse()
  if add:
    header_mutation.response.header_mutation.set_headers.extend(
        [
            service_pb2.HeaderValueOption(
                header=service_pb2.HeaderValue(
                    key=k, raw_value=bytes(v, "utf-8")
                )
            )
            for k, v in add
        ]
    )
  if remove is not None:
    header_mutation.response.header_mutation.remove_headers.extend(remove)
  if clear_route_cache:
    header_mutation.response.clear_route_cache = True
  return header_mutation


def add_body_mutation(
    body: str | None = None,
    clear_body: bool = False,
    clear_route_cache: bool = False,
) -> service_pb2.BodyResponse:
  """Generate a body response for incoming requests.

  Args:
    body: Text of the body.
    clear_body: If set to true, the modififcation will clear the previous body,
      if left false, the text will be appended to the end of the of the previous
      body.
    clear_route_cache: If true, will enable clear_route_cache on the response.

  Returns:
    The constructed body response object.
  """
  body_mutation = service_pb2.BodyResponse()
  if body:
    body_mutation.response.body_mutation.body = bytes(body, "utf-8")
  if clear_body:
    body_mutation.response.body_mutation.clear_body = True
  if clear_route_cache:
    body_mutation.response.clear_route_cache = True
  return body_mutation


class HealthCheckServer(BaseHTTPRequestHandler):
  """Server for responding to health check pings."""

  def do_GET(self) -> None:
    """Returns an empty page with 200 status code."""
    self.send_response(200)
    self.end_headers()


class CalloutServer(service_pb2_grpc.ExternalProcessorServicer):
  """Server for capturing and responding to service callout requests.

  Attributes:
    ip: Address that the main, server will attempt to connect to.
    port: Serving port of the main service.
    insecure_port: If using a grpc server, the port to serve non secure traffic on.
    health_check_ip: The health check serving address.
    health_check_port: Serving port of the health check service.
    server_thread_count: Threads allocated to the main grpc service.
    serperate_health_check: If set to false, will not bring up a the health
      check service.
    cert: If speficied, certificate used to authenticate the main grpc service
      for secure htps and http connections. If not specified will attempt to
      load data from a file pointed to by the cert_path.
    cert_path: Relative file path pointing to the main grpc certificate, cert.
    cert_key: Public key of the grpc certificate.
    cert_key_path: Relative file path pointing to the cert_key.
    root_cert: Root certificate for the main grpc service.
    root_cert_path: Relative file path pointing to the root_cert.
  """

  def __init__(
      self,
      ip: str = "0.0.0.0",
      port: int = 8443,
      insecure_port: int = 8080,
      health_check_ip: str = "0.0.0.0",
      health_check_port: int = 8000,
      serperate_health_check: bool = False,
      cert: bytes | None = None,
      cert_path: str = "../ssl_creds/localhost.crt",
      cert_key: bytes | None = None,
      cert_key_path: str = "../ssl_creds/localhost.key",
      root_cert: bytes | None = None,
      root_cert_path: str = "../ssl_creds/root.crt",
      server_thread_count: int = 2,
  ):
    self.ip = ip
    self.port = port
    self.insecure_port = insecure_port
    self.health_check_ip = health_check_ip
    self.health_check_port = health_check_port
    self.server_thread_count = server_thread_count
    self.serperate_health_check = serperate_health_check

    # read cert data
    if not cert:
      with open(cert_path, "rb") as file:
        self.cert = file.read()
        file.close()
    else:
      self.cert = cert

    if not cert_key:
      with open(cert_key_path, "rb") as file:
        self.cert_key = file.read()
        file.close()
    else:
      self.cert_key = cert_key

    if not root_cert:
      with open(root_cert_path, "rb") as file:
        self.root_cert = file.read()
        file.close()
    else:
      self.root_cert = root_cert

  def run(self):
    if not self.serperate_health_check:
      health_server = HTTPServer(
          (self.health_check_ip, self.health_check_port), HealthCheckServer
      )
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=self.server_thread_count)
    )
    service_pb2_grpc.add_ExternalProcessorServicer_to_server(self, server)
    server_credentials = grpc.ssl_server_credentials(
        private_key_certificate_chain_pairs=[(self.cert_key, self.cert)]
    )
    server.add_secure_port("%s:%d" % (self.ip, self.port), server_credentials)
    server.add_insecure_port("%s:%d" % (self.ip, self.insecure_port))
    server.start()
    print(
        "Server started, listening on %s:%d and %s:%d"
        % (self.ip, self.port, self.ip, self.insecure_port)
    )
    try:
      if not self.serperate_health_check:
        health_server.serve_forever()
    except KeyboardInterrupt:
      print("Server interrupted")
    finally:
      server.stop()
      if not serperate_health_check:
        health_server.server_close()

  def Process(
      self,
      request_iterator: Iterator[service_pb2.ProcessingRequest],
      context: ServicerContext,
  ) -> Iterator[service_pb2.ProcessingResponse]:
    """Process the client request."""
    for request in request_iterator:
      if request.HasField("request_headers"):
        yield service_pb2.ProcessingResponse(
            request_headers=self.on_request_headers(request.request_headers, context)
        )
      if request.HasField("response_headers"):
        yield service_pb2.ProcessingResponse(
            response_headers=self.on_response_headers(request.response_headers, context)
        )
      if request.HasField("request_body"):
        yield service_pb2.ProcessingResponse(
            request_body=self.on_request_body(request.request_body, context)
        )
      if request.HasField("response_body"):
        yield service_pb2.ProcessingResponse(
            response_body=self.on_response_body(request.response_body, context)
        )

  def on_request_headers(
      self, headers: service_pb2.HttpHeaders, context: ServicerContext
  ) -> service_pb2.HeadersResponse:
    """Process incoming request headers.

    Args:
      headers: Request headers to process.
      context: RPC context of the incoming request.

    Returns:
      Header modification object.
    """
    return None

  def on_response_headers(
      self, headers: service_pb2.HttpHeaders, context: ServicerContext
  ) -> service_pb2.HeadersResponse:
    """Process incoming response headers.

    Args:
      headers: Response headers to process.
      context: RPC context of the incoming request.

    Returns:
      Header modification object.
    """
    return None

  def on_request_body(
      self, body: service_pb2.HttpBody, context: ServicerContext
  ) -> service_pb2.BodyResponse:
    """Process an incoming request body.

    Args:
      headers: Request body to process.
      context: RPC context of the incoming request.

    Returns:
      Body modification object.
    """
    return None

  def on_response_body(
      self, body: service_pb2.HttpBody, context: ServicerContext
  ) -> service_pb2.BodyResponse:
    """Process an incoming response body.

    Args:
      headers: Response body to process.
      context: RPC context of the incoming request.

    Returns:
      Body modification object.
    """
    return None


if __name__ == "__main__":
  # Run the gRPC service
  CalloutServer().run()

File: callouts/java/src/main/java/ServiceCallout.java
/*
 * Copyright 2015 The gRPC Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import com.google.common.primitives.Bytes;
import com.google.protobuf.ByteString;
import io.envoyproxy.envoy.service.ext_proc.v3.BodyMutation;
import io.envoyproxy.envoy.service.ext_proc.v3.BodyResponse;
import io.envoyproxy.envoy.service.ext_proc.v3.CommonResponse;
import io.envoyproxy.envoy.service.ext_proc.v3.ExternalProcessorGrpc;
import io.envoyproxy.envoy.service.ext_proc.v3.HeaderMutation;
import io.envoyproxy.envoy.service.ext_proc.v3.HeaderValueOption;
import io.envoyproxy.envoy.service.ext_proc.v3.HeadersResponse;
import io.envoyproxy.envoy.service.ext_proc.v3.HttpBody;
import io.envoyproxy.envoy.service.ext_proc.v3.HttpHeaders;
import io.envoyproxy.envoy.service.ext_proc.v3.ProcessingRequest;
import io.envoyproxy.envoy.service.ext_proc.v3.ProcessingResponse;
import io.grpc.Grpc;
import io.grpc.Server;
import io.grpc.TlsServerCredentials;
import io.grpc.stub.StreamObserver;
import java.io.File;
import java.io.IOException;
import java.util.Map;
import java.util.Map.Entry;
import java.util.concurrent.TimeUnit;
import java.util.logging.Level;
import java.util.logging.Logger;

/** Server that manages startup/shutdown of a {@code Greeter} server. */
public class ServiceCallout {
  private static final Logger logger = Logger.getLogger(ServiceCallout.class.getName());

  private Server server;

  public ServiceCallout() {
    this(null, null, null, null, null, null, null, null, null, null, null, null);
  }

  public ServiceCallout(
      String ip,
      Integer port,
      Integer insecurePort,
      String healthCheckIp,
      Integer healthCheckPort,
      Boolean serperateHealthCheck,
      Bytes cert,
      String certPath,
      Bytes certKey,
      String certKeyPath,
      Integer serverThreadCount,
      Boolean enableInsecurePort) {
    if (ip != null) {
      this.ip = ip;
    }
    if (port != null) {
      this.port = port;
    }
    if (insecurePort != null) {
      this.insecurePort = insecurePort;
    }
    if (healthCheckIp != null) {
      this.healthCheckIp = healthCheckIp;
    }
    if (healthCheckPort != null) {
      this.healthCheckPort = healthCheckPort;
    }
    if (serperateHealthCheck != null) {
      this.serperateHealthCheck = serperateHealthCheck;
    }
    if (cert != null) {
      this.cert = cert;
    }
    if (certPath != null) {
      this.certPath = certPath;
    }
    if (certKey != null) {
      this.certKey = certKey;
    }
    if (certKeyPath != null) {
      this.certKeyPath = certKeyPath;
    }
    if (serverThreadCount != null) {
      this.serverThreadCount = serverThreadCount;
    }
    if (enableInsecurePort != null) {
      this.enableInsecurePort = enableInsecurePort;
    }
  }

  private String ip = "0.0.0.0";
  private int port = 8443;
  private int insecurePort = 8443;
  private String healthCheckIp = "0.0.0.0";
  private int healthCheckPort = 8000;
  private boolean serperateHealthCheck = false;
  private Bytes cert = null;
  private String certPath = "../ssl_creds/localhost_public.pem";
  private Bytes certKey = null;
  private String certKeyPath = "../ssl_creds/localhost_private.pem";
  private int serverThreadCount = 2;
  private boolean enableInsecurePort = true;

  private Server healthCheckServer;

  // private void startHealthCheckServer() throws IOException {
  //   if (!serperateHealthCheck) {
  //     HealthStatusManager healthStatusManager = new HealthStatusManager();
  //     BindableService healthService = healthStatusManager.getHealthService();
  //     healthCheckServer =
  //         ServerBuilder.forPort(healthCheckPort).addService(healthService).build().start();
  //     healthStatusManager.setStatus("", ServingStatus.SERVING);
  //   }
  // }

  private void stopHealthCheckServer() throws InterruptedException {
    if (!serperateHealthCheck) {
      healthCheckServer.shutdown().awaitTermination(30, TimeUnit.SECONDS);
    }
  }

  public void start() throws IOException {
    /* The port on which the server should run */
    TlsServerCredentials.Builder tlsBuilder =
        TlsServerCredentials.newBuilder().keyManager(new File(certPath), new File(certKeyPath));

    // Setup the callout extension service.
    server =
        Grpc.newServerBuilderForPort(port, tlsBuilder.build())
            .addService(new ExternalProcessorImpl())
            .build()
            .start();

    // // Setup the health check service.
    // startHealthCheckServer();

    // Configure callout shutdown hook.
    logger.info("Server started, listening on " + port);
    Runtime.getRuntime()
        .addShutdownHook(
            new Thread() {
              @Override
              public void run() {
                // Use stderr here since the logger may have been reset by its JVM shutdown hook.
                System.err.println("*** shutting down gRPC server since JVM is shutting down");
                try {
                  stopHealthCheckServer();
                  ServiceCallout.this.stop();
                } catch (InterruptedException e) {
                  e.printStackTrace(System.err);
                }
                System.err.println("*** server shut down");
              }
            });
  }

  public void stop() throws InterruptedException {
    if (server != null) {
      server.shutdown().awaitTermination(30, TimeUnit.SECONDS);
    }
  }

  /** Await termination on the main thread since the grpc library uses daemon threads. */
  public void blockUntilShutdown() throws InterruptedException {
    if (server != null) {
      server.awaitTermination();
    }
  }

  public void AddHeaderMutations(
      HeadersResponse.Builder headersResponseBuilder, Iterable<Map.Entry<String, String>> add) {
    if (add != null) {
      HeaderMutation.Builder headerMutationBuilder =
          headersResponseBuilder.getResponseBuilder().getHeaderMutationBuilder();
      for (Entry<String, String> entry : add) {
        headerMutationBuilder
            .addSetHeadersBuilder()
            .getHeaderBuilder()
            .setKey(entry.getKey())
            .setRawValue(ByteString.copyFromUtf8(entry.getValue()));
      }
    }
  }

  public HeadersResponse ConfigureHeadersResponse(
      HeadersResponse.Builder headersResponseBuilder,
      Iterable<HeaderValueOption> add,
      Iterable<String> remove,
      Boolean clearRouteCache) {
    CommonResponse.Builder responseBuilder = headersResponseBuilder.getResponseBuilder();
    HeaderMutation.Builder headerBuilder = responseBuilder.getHeaderMutationBuilder();
    if (add != null) {
      headerBuilder.addAllSetHeaders(add);
    }
    if (remove != null) {
      headerBuilder.addAllRemoveHeaders(remove);
    }
    if (clearRouteCache != null) {
      responseBuilder.setClearRouteCache(clearRouteCache);
    }
    return headersResponseBuilder.build();
  }

  public BodyResponse BuildBodyMutationResponse(
      BodyResponse.Builder bodyResponseBuilder,
      String body,
      Boolean clearBody,
      Boolean clearRouteCache) {
    CommonResponse.Builder responseBuilder = bodyResponseBuilder.getResponseBuilder();
    BodyMutation.Builder bodyBuilder = responseBuilder.getBodyMutationBuilder();
    if (body != null) {
      bodyBuilder.setBody(ByteString.copyFromUtf8(body));
    }
    if (clearBody != null) {
      bodyBuilder.setClearBody(clearBody);
    }
    if (clearRouteCache != null) {
      responseBuilder.setClearRouteCache(clearRouteCache);
    }
    return bodyResponseBuilder.build();
  }

  public ProcessingResponse ProcessRequest(ProcessingRequest request) {
    ProcessingResponse.Builder builder = ProcessingResponse.newBuilder();
    switch (request.getRequestCase()) {
      case REQUEST_HEADERS:
        OnRequestHeaders(builder.getRequestHeadersBuilder(), request.getRequestHeaders());
        break;
      case RESPONSE_HEADERS:
        OnResponseHeaders(builder.getRequestHeadersBuilder(), request.getResponseHeaders());
        break;
      case REQUEST_BODY:
        OnRequestBody(builder.getRequestBodyBuilder(), request.getRequestBody());
        break;
      case RESPONSE_BODY:
        OnResponseBody(builder.getResponseBodyBuilder(), request.getResponseBody());
        break;
      case REQUEST_TRAILERS:
        break;
      case RESPONSE_TRAILERS:
        break;
      case REQUEST_NOT_SET:
      default:
        logger.log(Level.WARNING, "Receieved a ProcessingRequest with no request data.");
        break;
    }
    return builder.build();
  }

  public void OnRequestHeaders(HeadersResponse.Builder headerResponse, HttpHeaders headers) {
    logger.log(Level.INFO, "Unhandled RequestHeader recieved, returning an empty HeadersResponse.");
  }

  public void OnResponseHeaders(HeadersResponse.Builder headerResponse, HttpHeaders headers) {
    logger.log(
        Level.INFO, "Unhandled ResponseHeaders recieved, returning an empty HeadersResponse.");
  }

  public void OnRequestBody(BodyResponse.Builder bodyResponse, HttpBody body) {
    logger.log(Level.INFO, "Unhandled RequestBody recieved, returning an empty BodyResponse.");
  }

  public void OnResponseBody(BodyResponse.Builder bodyResponse, HttpBody body) {
    logger.log(Level.INFO, "Unhandled ResponseBody recieved, returning an empty BodyResponse.");
  }

  /** Main launches the server from the command line. */
  public static void main(String[] args) throws IOException, InterruptedException {
    final ServiceCallout server = new ServiceCallout();
    server.start();
    server.blockUntilShutdown();
  }

  private class ExternalProcessorImpl extends ExternalProcessorGrpc.ExternalProcessorImplBase {

    @Override
    public StreamObserver<ProcessingRequest> process(
        final StreamObserver<ProcessingResponse> responseObserver) {
      // responseObserver.onCompleted();
      return new StreamObserver<ProcessingRequest>() {
        @Override
        public void onNext(ProcessingRequest request) {
          responseObserver.onNext(ProcessRequest(request));
        }

        @Override
        public void onError(Throwable t) {
          logger.log(Level.WARNING, "Encountered request error", t);
        }

        @Override
        public void onCompleted() {
          responseObserver.onCompleted();
        }
      };
    }
  }
}
